ingestion/src/metadata/ingestion/lineage/parser.py

```diff
@@ -37,214 +43,283 @@
 logger = ingestion_logger()
 
 
-def get_involved_tables_from_parser(
-    parser: LineageRunner,
-) -> Optional[List[models.Table]]:
-    """
-    Use the LineageRunner parser and combine
-    source and intermediate tables into
-    a single set.
-    :param parser: LineageRunner
-    :return: List of involved tables
-    """
-    try:
+class LineageParser:
+
+    parser: LineageRunner
+    query: str
+    _clean_query: str
+
+    def __init__(self, query: str):
+        self.query = query
+        self._clean_query = self.clean_raw_query(query)
+        self.parser = LineageRunner(query)
+
+    @cached_property
+    def involved_tables(self) -> Optional[List[Table]]:
+        """
+        Use the LineageRunner parser and combine
+        source and intermediate tables into
+        a single set.
+        :return: List of involved tables
+        """
+        try:
+            return list(
+                set(self.source_tables)
+                .union(set(self.intermediate_tables))
+                .union(set(self.target_tables))
+            )
+        except SQLLineageException as exc:
+            logger.debug(traceback.format_exc())
+            logger.warning(
+                f"Cannot extract source table information from query [{self.query}]: {exc}"
+            )
+            return None
+
+    @cached_property
+    def intermediate_tables(self) -> List[Table]:
+        """
+        Get a list of intermediate tables
+        """
+        # These are @lazy_property, not properly being picked up by IDEs. Ignore the warning
+        return self.retrieve_tables(self.parser.intermediate_tables)
+
+    @cached_property
+    def source_tables(self) -> List[Table]:
+        """
+        Get a list of source tables
+        """
+        # These are @lazy_property, not properly being picked up by IDEs. Ignore the warning
+        return self.retrieve_tables(self.parser.source_tables)
+
+    @cached_property
+    def target_tables(self) -> List[Table]:
+        """
+        Get a list of target tables
+        """
         # These are @lazy_property, not properly being picked up by IDEs. Ignore the warning
-        return list(
-            set(parser.source_tables)
-            .union(set(parser.intermediate_tables))
-            .union(set(parser.target_tables))
+        return self.retrieve_tables(self.parser.target_tables)
+
+    @cached_property
+    def column_lineage(self) -> List[Tuple[Column, Column]]:
+        """
+        Get a list of tuples of column lineage
+        """
+        return self.parser.get_column_lineage()
+
+    @cached_property
+    def clean_table_list(self) -> List[str]:
+        """
+        Clean the table name if it has <default>.
+        :return: clean table names
+        """
+        return [get_formatted_entity_name(str(table)) for table in self.involved_tables]
+
+    @cached_property
```

Comment:
as discussed, this is quite beautiful but not supported on 3.7, so we'll need to manually handle the caching around here https://docs.python.org/3/library/functools.html#functools.cached_property

Reply:
Added a library to support it on Python 3.7